### PR TITLE
[gui] fixes #1388 Add PeerHeight field in /network/connection(s) endpoints

### LINKED ISSUE #1388: Include peer's reported block height in /connections responses
After this is done, then the client can display the value in the connections panel.

This data is already in /blockchain/progress but should be added to /connections data

### PR DESCRIPTION
**Pull request by olemis**
Fixes #1388 . Supersedes #1417 and #1447 .

Changes:

- Add `height` field in response of /network/connection and /network/connections API endpoints

Does this change need to mentioned in CHANGELOG.md?
yes

### CODE REVIEW

File: src/gui/network_test.go [hunk 1]
```diff
@@ -156,6 +203,7 @@ func TestConnections(t *testing.T) {
 			endpoint := "/network/connections"
 			gateway := NewGatewayerMock()
 			gateway.On("GetConnections").Return(tc.gatewayGetConnectionsResult)
+			gateway.On("GetBlockchainProgress").Return(&bp)
```
Comment (gz-c) [inline src/gui/network_test.go]: `bp` should be should be part of the test data struct

File: src/gui/network_test.go [hunk 2]
```diff
@@ -71,6 +94,7 @@ func TestConnection(t *testing.T) {
 			endpoint := "/network/connection"
 			gateway := NewGatewayerMock()
 			gateway.On("GetConnection", tc.addr).Return(tc.gatewayGetConnectionResult)
+			gateway.On("GetBlockchainProgress").Return(&bp)
```
Comment (gz-c) [inline src/gui/network_test.go]: `bp` should be should be part of the test data struct

File: src/gui/network.go [hunk 3]
```diff
@@ -26,19 +39,52 @@ func connectionHandler(gateway Gatewayer) http.HandlerFunc {
...
 		}
 
-		wh.SendJSONOr500(logger, w, gateway.GetConnections())
+		dcnxs := gateway.GetConnections()
+		bcp := gateway.GetBlockchainProgress()
+		index := peerHeightIndex(bcp)
+		cnxs := Connections{}
+		for _, c := range dcnxs.Connections {
+			cnx := Connection{c, 0}
+			if height, hasRecord := index[c.Addr]; hasRecord {
```
Comment (gz-c) [inline src/gui/network.go]: height would be 0 if not found in the map, so it is not necessary to check `hasRecord`

File: src/gui/network.go [hunk 4]
```diff
@@ -26,19 +39,52 @@ func connectionHandler(gateway Gatewayer) http.HandlerFunc {
...
 			return
 		}
 
-		wh.SendJSONOr500(logger, w, gateway.GetConnections())
+		dcnxs := gateway.GetConnections()
+		bcp := gateway.GetBlockchainProgress()
+		index := peerHeightIndex(bcp)
+		cnxs := Connections{}
+		for _, c := range dcnxs.Connections {
+			cnx := Connection{c, 0}
```
Comment (gz-c) [inline src/gui/network.go]: name initializers are preferred, e.g.

```go
Connection{
    Connection: c,
    Height: index[c.Addr],
}
```

File: src/gui/network.go [hunk 5]
```diff
@@ -26,19 +39,52 @@ func connectionHandler(gateway Gatewayer) http.HandlerFunc {
...
+			}
+			i++
+		}
 
-		wh.SendJSONOr500(logger, w, c)
+		wh.SendJSONOr500(logger, w, cnx)
 	}
 }
 
+func peerHeightIndex(bcp *daemon.BlockchainProgress) (index map[string]uint64) {
```
Comment (gz-c) [inline src/gui/network.go]: this can be inlined in the method that is using it

File: src/gui/network.go [hunk 6]
```diff
@@ -26,19 +39,52 @@ func connectionHandler(gateway Gatewayer) http.HandlerFunc {
...
 		}
+		cnx := Connection{c, 0}
+		bcp := gateway.GetBlockchainProgress()
+		notFound := true
+		i, l := 0, len(bcp.Peers)
+		for notFound && i < l {
+			ph := bcp.Peers[i]
+			if ph.Address == c.Addr {
+				notFound = false
+				cnx.Height = ph.Height
```
Comment (gz-c) [inline src/gui/network.go]:
```go
for _, ph := range bcp.Peers {
    if ph.Address == cnx.Addr {
        cnx.Height = ph.Height
        break
    }
}
```

File: src/gui/network.go [hunk 7]
```diff
@@ -5,9 +5,22 @@ import (
...
+// Connection wrapper around daemon connection with info about block height added
+type Connection struct {
+	*daemon.Connection
+	Height uint64 `json:"height"`
+}
+
+// Connections an array of connections
+// Arrays must be wrapped in structs to avoid certain javascript exploits
+type Connections struct {
+	Connections []*Connection `json:"connections"`
```
[the remark this answers is not in the document]
Comment (olemis) [reply src/gui/network.go]: This is the same as `daemon.Connections` struct . I just kept the same approach . No major difference in the end though .

[the remark this answers is not in the document]
Comment (olemis) [reply src/gui/network.go]: done

File: src/api/network.go [hunk 8]
```diff
@@ -26,8 +39,23 @@ func connectionHandler(gateway Gatewayer) http.HandlerFunc {
 			wh.Error404(w, "")
 			return
 		}
+		cnx := Connection{
+			Connection: c,
+			Height:     0,
+		}
+		bcp, err := gateway.GetBlockchainProgress()
+		if err != nil {
+			wh.Error500(w, "GetBlockchainProgress failed")
```
Comment (gz-c) [inline src/api/network.go]: should write the error message

Comment (gz-c) [inline src/api/network.go]: should write the error message

File: src/api/network.go [hunk 9]
```diff
@@ -38,7 +66,29 @@ func connectionsHandler(gateway Gatewayer) http.HandlerFunc {
 			return
 		}
 
-		wh.SendJSONOr500(logger, w, gateway.GetConnections())
+		dcnxs := gateway.GetConnections()
+		bcp, err := gateway.GetBlockchainProgress()
+		if err != nil {
+			wh.Error500(w, "GetBlockchainProgress failed")
```
Comment (gz-c) [inline src/api/network.go]: should write the error message

File: src/api/network_test.go [hunk 10]
```diff
@@ -14,14 +14,41 @@ import (
...
-		addr                       string
-		gatewayGetConnectionResult *daemon.Connection
-		result                     *daemon.Connection
+		name                               string
+		method                             string
+		status                             int
+		err                                string
+		addr                               string
+		gatewayGetConnectionResult         *daemon.Connection
+		gatewayGetBlockchainProgressResult *daemon.BlockchainProgress
```
Comment (gz-c) [inline src/api/network_test.go]: should have test that mocks GetBlockchainProgress with an error

File: src/api/network_test.go [hunk 11]
```diff
@@ -102,13 +131,40 @@ func TestConnection(t *testing.T) {
...
-		status                      int
-		err                         string
-		gatewayGetConnectionsResult *daemon.Connections
-		result                      *daemon.Connections
+		name                               string
+		method                             string
+		status                             int
+		err                                string
+		gatewayGetConnectionsResult        *daemon.Connections
+		gatewayGetBlockchainProgressResult *daemon.BlockchainProgress
```
Comment (gz-c) [inline src/api/network_test.go]: should have test that mocks GetBlockchainProgress with an error

File: src/api/network_test.go (hunk 10 above)
Comment (olemis) [reply src/api/network_test.go]: done

File: src/api/network_test.go (hunk 11 above)
Comment (olemis) [reply src/api/network_test.go]: done

File: src/api/network_test.go (hunk 10 above)
Comment (gz-c) [reply src/api/network_test.go]: still need to write the original error message, not replaced with "GetBlockchainProgress failed"

Comment (olemis) [reply src/api/network_test.go]: done

**Pull request merged**